### PR TITLE
Deprecate casting to `unmanaged`

### PR DESCRIPTION
**Pull request by jabraham17**
Based on offline discussion, we decided to deprecate casting to `unmanaged` from other managed types (`owned`, `shared`, `borrowed`)

## Summary of changes
- deprecates casting to `unmanaged`
- fixes tests and modules which rely on the behavior
  - In some cases, removed tests which solely tested the deprecated behavior

## New tests
- add deprecation test

## testing
- paratest + futures
- paratest + gasnet
- paratest + memleaks
- paratest + gasnet + memleaks
- docs testing

***

[Reviewed by @vasslitvinov]

Based on offline discussion in cray/chapel-private#4755
Closes cray/chapel-private#4821

### CODE REVIEW

File: modules/internal/ChapelBase.chpl [hunk 1]
```diff
@@ -2150,23 +2150,45 @@ module ChapelBase {
   }
 
   // casting to unmanaged?, no class downcast
+  pragma "last resort"
+  @deprecated("casting from a managed class to an 'unmanaged?' is deprecated")
```
Comment (vasslitvinov) [inline modules/internal/ChapelBase.chpl:2154]: "borrowed" is not a "managed" category, if I understand correctly. At the very least the deprecation text should be different.

**Comment by vasslitvinov** [comment]
Chapel provides "unmanaged" to enable users to bypass the management controls in those cases where these controls are overly restrictive. "unmanaged" is inherently unsafe, so a cast to "unmanaged" does not make the situation any worse. For this reason I am skeptical about the value of this deprecation.

At the very least, when a method is invoked on an "unmanaged", users should be able to access the unmanaged `this` inside the method body. Today the only way to do so is to cast to "unmanaged"' in the method body. If we disallow this cast, users will lose unmanaged-ness of the receiver. And of course if we allow this cast, then a user will be able to cast owned and shared to unmanaged by going through a borrow.

**Comment by jabraham17** [comment]
> when a method is invoked on an "unmanaged", users should be able to access the unmanaged this inside the method body

While doing the deprecation, this was the primary use case. This was mainly done in internal modules where `_to_unmanaged()` was used as a replacement. The other uses where easily worked around without requiring casts.

I would argue this is a case where we could turn `_to_unmanaged()` into a user facing function, particularly if there are use cases.

File: modules/internal/ChapelBase.chpl (hunk 1 above)
Comment (jabraham17) [reply modules/internal/ChapelBase.chpl:2154]: I am also unhappy with the deprecation message. But due to how the compiler resolves casts (compiler/resolution/functionResolution.cpp:3087), I could not find a way to get the true type being casted from, it always returned `borrowed`. I decided that the efforts to make it work were not worth it for a 6 month deprecation. I am definitely open to better wordsmithing here though

**Review by vasslitvinov — APPROVED** [review]
Looks good.

**Pull request merged**